actor/kvcounter/src/lib.rs

```diff
@@ -0,0 +1,46 @@
+use serde_json::json;
+use wasmbus_rpc::actor::prelude::*;
+use wasmcloud_interface_httpserver::{HttpRequest, HttpResponse, HttpServer, HttpServerReceiver};
+use wasmcloud_interface_keyvalue::{IncrementRequest, KeyValue, KeyValueSender};
+
+#[derive(Debug, Default, Actor, HealthResponder)]
+#[services(Actor, HttpServer)]
+struct KvCounterActor {}
+
+/// Implementation of HttpServer trait methods
+#[async_trait]
+impl HttpServer for KvCounterActor {
+    async fn handle_request(&self, ctx: &Context, req: &HttpRequest) -> RpcResult<HttpResponse> {
+        // make friendlier key
+        let key = format!("counter:{}", req.path.replace('/', ":"));
+
+        // bonus: use specified amount from query, or 1
+        let amount: i32 = form_urlencoded::parse(req.query_string.as_bytes())
+            .find(|(n, _)| n == "amount")
+            .map(|(_, v)| v.parse::<i32>())
+            .unwrap_or(Ok(1))
+            .unwrap_or(1);
+
+        // increment the value in kv and send response in json
+        let (body, status_code) = match increment_counter(ctx, key, amount).await {
+            Ok(v) => (json!({ "counter": v }).to_string(), 200),
+            // if we caught an error, return it to client
+            Err(e) => (json!({ "error": e.to_string() }).to_string(), 500),
+        };
+        let resp = HttpResponse {
+            body: body.as_bytes().to_vec(),
+            status_code,
+            ..Default::default()
+        };
+        Ok(resp)
+    }
+}
+
+/// increment the counter by the amount, returning the new value
+async fn increment_counter(ctx: &Context, key: String, value: i32) -> RpcResult<i32> {
+    let tx = WasmHost::to_provider("wasmcloud:keyvalue", "default")?;
+    let kv = KeyValueSender::new(&tx);
+
+    let new_val = kv.increment(ctx, &IncrementRequest { key, value }).await?;
+    Ok(new_val as i32)
+}
```

Comment:
I've been looking at this block of code for a little while as the steps feel like they could be consolidated a bit, and cleaned up for the user. There's nothing that needs to be done now, but I have a few questions / suggestions
1. Could we skip the `tx` step where we create a `WasmHost` and simply give the `KeyValueSender::new` a contract  ID and link definition?
2. Along the same lines, could we pass the context into the `KeyValueSender::new` constructor or is context supposed to be operation specific (e.g. is there a use case for multiple contexts for multiple KeyValueSenders`

If both of the above optimizations could be done, that would change this code to be:
```
/// increment the counter by the amount, returning the new value
async fn increment_counter(ctx: &Context, key: String, value: i32) -> RpcResult<i32> {
    let kv = KeyValueSender::new(ctx, "wasmcloud:keyvalue", "default")?;
    let new_val = kv.increment(&IncrementRequest { key, value }).await?;
    Ok(new_val as i32)
}
``` 

It's not much of a difference here, but from what I see it looks like these steps could certainly be consolidated. I also dislike always having `ctx` as the first parameter but I'll admit this is a common pattern and I haven't used it much.

For this PR, I don't feel like this has to be done so this is mostly comment/question material, want to make sure I'm not making assumptions

Reply:
Those are good questions which I will address in a different forum, since they are longish-answers and I'm sure others will ask about it later. Here's the TL;DR:

The short answer to the first one is that we could consolidate those two lines and save a line of code. Since it's only one line of code for the actor developer, I don't think it rises to a requirement for MVP.

ctx is in part a hook to incorporate environmental or other data that you might need, and it doesn't create a breaking api change if you add those later.

Putting Context into the constructor would save the ctx parameter on ServiceXSender api calls, but we still need it on the ServiceXReceiver interfaces, because providers need it to know the actor that sent the call. Right now, both *Sender and *Receiver implement the same ServiceX trait. It would be possible to generate different traits for sender and receiver, but even then, actors that act as receiver for a trait would still see a ctx parameter on every handler. I think it might still be a good idea to experiment with, to see what the resulting apis look like. This would be a big change to codegen, and would cause some destabilization if we did it before 0.50.

Reply:
Thank you for the TL;DR, agreed that the little optimizations aren't a requirement for MVP. Definitely want to avoid premature optimizations that may require us to do something like adding in the `ctx` parameter later on. I think we'll know better from the feedback we get from people who play with the new interfaces / actor API what we need to improve on